pkg/deploy/broker.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/submariner-io/submariner-operator/pkg/subctl/components"
+	v1 "k8s.io/api/core/v1"
+
+	submarinerv1a1 "github.com/submariner-io/submariner-operator/api/submariner/v1alpha1"
+	"github.com/submariner-io/submariner-operator/pkg/internal/cli"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/datafile"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/operator/brokercr"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/operator/submarinerop"
+)
+
+type DeployOptions struct {
+	OperatorDebug               bool
+	GlobalnetEnable             bool
+	IpsecSubmFile               string
+	GlobalnetCIDRRange          string
+	DefaultGlobalnetClusterSize uint
+	ComponentArr                []string
+	GlobalCIDRConfigMap         *v1.ConfigMap
+	DefaultCustomDomains        []string
+	Repository                  string
+	ImageVersion                string
+	BrokerNamespace             string
+}
+
+var ValidComponents = []string{components.ServiceDiscovery, components.Connectivity}
+
+const brokerDetailsFilename = "broker-info.subm"
+
+func Broker(do DeployOptions, kubeConfig, kubeContext string) error {
+	status := cli.NewStatus()
+	componentSet := stringset.New(do.ComponentArr...)
+
+	if err := isValidComponents(componentSet); err != nil {
+		return fmt.Errorf("invalid components parameter %s", err)
+	}
+
+	if do.GlobalnetEnable {
+		componentSet.Add(components.Globalnet)
+	}
+
+	if valid, err := isValidGlobalnetConfig(do); !valid {
+		if err != nil {
+			return fmt.Errorf("invalid GlobalCIDR configuration %s", err)
+		}
+	}
+
+	config, err := restconfig.ForCluster(kubeConfig, kubeContext)
+	if err != nil {
+		return fmt.Errorf("the provided kubeconfig is invalid %s", err)
+	}
+
+	status.Start("Setting up broker RBAC")
+	err = broker.Ensure(config, do.ComponentArr, false, do.BrokerNamespace)
+	status.End(cli.CheckForError(err))
+	if err != nil {
+		return fmt.Errorf("error setting up broker RBAC %s", err)
+	}
+
+	status.Start("Deploying the Submariner operator")
+	err = submarinerop.Ensure(status, config, internal.OperatorNamespace,
+		image.Operator(do.ImageVersion, do.Repository, nil), do.OperatorDebug)
+	status.End(cli.CheckForError(err))
+	if err != nil {
+		return fmt.Errorf("error deploying the operator %s", err)
+	}
+
+	status.Start("Deploying the broker")
+	err = brokercr.Ensure(config, do.BrokerNamespace, populateBrokerSpec(do))
+	if err == nil {
+		status.QueueSuccessMessage("The broker has been deployed")
+		status.End(cli.Success)
+	} else {
+		status.QueueFailureMessage("Broker deployment failed")
+		status.End(cli.Failure)
+		return fmt.Errorf("error deploying the broker %s", err)
+	}
+
+	status.Start(fmt.Sprintf("Creating %s file", brokerDetailsFilename))
+
+	// If deploy-broker is retried we will attempt to re-use the existing IPsec PSK secret
+	if do.IpsecSubmFile == "" {
+		if _, err := datafile.NewFromFile(brokerDetailsFilename); err == nil {
```

Comment:
This also doesn’t work in a library context, there is no guarantee that the caller will be using files to store the data.

Reply:
Yes. I will make those changes as part of another PR.